File: aoc_2025_rust/10/src/bin/part2.py
from z3 import Optimize, Int, If, sat, Sum


def get_input(filename):
    result = []
    with open(filename, "r") as f:
        for line in f:
            s = line.strip()
            if not s:
                continue

            first = s.find(' ')
            last = s.rfind(' ')

            part2 = s[first + 1:last]
            part3 = s[last + 1:]

            part2_list = []
            for tok in part2.split():
                tok_clean = tok.strip("()")
                nums = [int(x) for x in tok_clean.split(',') if x]
                part2_list.append(nums)
            part3_list = [int(x) for x in part3.replace('{', '').replace('}', '').split(',') if x.strip()]

            result.append((part2_list, part3_list))

    return result


def solve_min_button_presses(buttons, target):
    n = len(target)
    b = len(buttons)

    opt = Optimize()

    count = [Int(f"c_{i}") for i in range(b)]
    for c in count:
        opt.add(c >= 0)  # cant press a button negative times

    for i in range(n):
        expr = Sum([If(i in buttons[btn], count[btn], 0) for btn in range(b)])
        opt.add(expr == target[i])

    opt.minimize(Sum(count))

    if opt.check() == sat:
        model = opt.model()
        button_counts = [model.eval(c).as_long() for c in count]
        sequence = []
        for btn, cnt in enumerate(button_counts):
            sequence.extend([btn] * cnt)

        return sequence
    else:
        return None


if __name__ == "__main__":
    input_data = get_input("src/input.txt")

    total = 0
    for i, (buttons, target) in enumerate(input_data):
        seq = solve_min_button_presses(buttons, target)
        total += len(seq)

    print(total)

File: aoc_2025_rust/10/src/bin/part2.rs
use day_10::utils::get_input;

fn main() {
    let input = get_input("src/test_input.txt");
    let mut total = 0;

    fn press(state: &Vec<usize>, button: &Vec<usize>) -> Vec<usize> {
        let mut output = state.clone();

        for idx in button.iter() {
            output[*idx] += 1;
        }

        output
    }

    'main: for (_, buttons, target) in input.iter() {
        let mut queue: Vec<(Vec<usize>, Vec<usize>)> = Vec::new();
        queue.push((vec![0; target.len()], Vec::<usize>::new()));
        
        loop {
            let mut new_queue: Vec<(Vec<usize>, Vec<usize>)> = Vec::new();
            for (state, pressed) in queue.iter() {
                'inner: for (button_idx, button) in buttons.iter().enumerate() {
                    let new_state = press(&state, &button);
                    let mut new_pressed = pressed.clone();
                    new_pressed.push(button_idx);

                    for (cur_val, target_val) in new_state.iter().zip(target.iter()) {
                        if cur_val > target_val {
                            continue 'inner;
                        }
                    }
    
                    if new_state == *target {
                        total += new_pressed.len();
                        continue 'main;
                    }
    
                    new_queue.push((new_state.clone(), new_pressed));
                }
            }
            queue = new_queue;
        }
    }

    println!("{}", total);
}